src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs

```diff
@@ -118,19 +115,25 @@ private object VisitCache(ServiceCallSite callSite, RuntimeResolverContext conte
 
         private object ResolveService(ServiceCallSite callSite, RuntimeResolverContext context, RuntimeResolverLock lockType, ServiceProviderEngineScope serviceProviderEngine)
         {
+            object resolved;
             Dictionary<ServiceCacheKey, object> resolvedServices = serviceProviderEngine.ResolvedServices;
-            if (!resolvedServices.TryGetValue(callSite.Cache.Key, out object resolved))
+            lock (resolvedServices)
             {
-                resolved = VisitCallSiteMain(callSite, new RuntimeResolverContext
+                if (resolvedServices.TryGetValue(callSite.Cache.Key, out resolved))
                 {
-                    Scope = serviceProviderEngine,
-                    AcquiredLocks = context.AcquiredLocks | lockType
-                });
-
-                serviceProviderEngine.CaptureDisposable(resolved);
+                    return resolved;
+                }
+            }
+            resolved = VisitCallSiteMain(callSite, new RuntimeResolverContext
```

Comment:
Add a comment here about not needing to lock here because of the lock taken outside of this method.